### PR TITLE
netpol: Use kube-router as a library 

### PR DESCRIPTION
**Pull request by vadorovsky**
#### Proposed Changes ####

Before this change, we were copying a part of kube-router code to
pkg/agent/netpol directory with modifications, from which the biggest
one was consumption of k3s node config instead of kube-router config.

However, that approach made it hard to follow new upstream versions.
It's possible to use kube-router as a library, so it seems like a better
way to do that.

Instead of modifying kube-router network policy controller to comsume
k3s configuration, this change just converts k3s node config into
kube-router config. All the functionality of kube-router except netpol
is still disabled.

#### Types of Changes ####

Refactoring

#### Verification ####

So far I did a simple network policy test according to the following docs:
https://kubernetes.io/docs/tasks/administer-cluster/declare-network-policy/

What I did was:
* deploying k3s on Vagrant environment
* creating an nginx deployment, exposing it as a service
* checking if I can connect to the service from a busybox pod - I could, as expected
* creating a simple network policy limiting access to nginx
* checking again if I can connect to the service from a busybox pod - I could **not**, as expected

```bash
root@k3s-1-ubuntu2104:~# ./dist/artifacts/k3s kubectl get svc,pod
NAME                 TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)   AGE
service/kubernetes   ClusterIP   10.43.0.1       <none>        443/TCP   22m
service/nginx        ClusterIP   10.43.160.204   <none>        80/TCP    6m13s

NAME                         READY   STATUS    RESTARTS   AGE
pod/nginx-6799fc88d8-k7vqz   1/1     Running   0          14m

root@k3s-1-ubuntu2104:~# ./dist/artifacts/k3s kubectl run busybox --rm -ti --image=busybox -- /bin/sh
If you don't see a command prompt, try pressing enter.
/ # wget --spider --timeout=1 nginx
Connecting to nginx (10.43.160.204:80)
remote file exists
/ # 
Session ended, resume using 'kubectl attach busybox -c busybox -i -t' command when the pod is running
pod "busybox" deleted

root@k3s-1-ubuntu2104:~# cat netpol.yaml 
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: access-nginx
spec:
  podSelector:
    matchLabels:
      app: nginx
  ingress:
  - from:
    - podSelector:
        matchLabels:
          access: "true"

root@k3s-1-ubuntu2104:~# ./dist/artifacts/k3s kubectl apply -f netpol.yaml
networkpolicy.networking.k8s.io/access-nginx created

root@k3s-1-ubuntu2104:~# ./dist/artifacts/k3s kubectl run busybox --rm -ti --image=busybox -- /bin/sh
If you don't see a command prompt, try pressing enter.
/ # wget --spider --timeout=1 nginx
Connecting to nginx (10.43.160.204:80)
wget: download timed out
```

#### User-Facing Change ####

```release-note
NONE
```

#### Further Comments ####

We need to be careful with testing it. With my simple test it works, but:
* I was not present in k3s community when @erikwilson and @brandond were implementing the kube-router fork, I'm maybe missing some info
* we need more tests to ensure that it really works for all scenarios

### CODE REVIEW

File: pkg/agent/run.go [hunk 1]
```diff
@@ -116,13 +122,19 @@ func run(ctx context.Context, cfg cmds.Agent, proxy proxy.Proxy) error {
...
 		}
 	}
 
 	if !nodeConfig.AgentConfig.DisableNPC {
-		if err := netpol.Run(ctx, nodeConfig); err != nil {
+		stopChan := make(chan struct{})
+		healthChan := make(chan *healthcheck.ControllerHeartbeat, healthControllerChannelLength)
+		defer func() {
+			close(stopChan)
+			close(healthChan)
```
Comment (vadorovsky) [inline pkg/agent/run.go:135]: I was a bit unsure about how to deal with those channnels. Basically, kube-router requires both of them for running all its controllers. It also requires a `WaitGroup`

In the other parts of k3s agent, we were not using any channels or WaitGroups, we were just relying in a `Context`. I was wondering - should we inroduce channels and WaitGroups everywhere? Modify kube-router to not require channels?

So far I decided to:
* keep the healtcheck server of kube-router, of course enable only monitoring of netpol controller
* keep the channels only for kube-router
* use the WaitGroup also for the flannel goroutime, because why not :man_shrugging: after all, the purpose of WaitGroups is to wait for all goroutines

WDYT?

File: pkg/agent/run.go [hunk 2]
```diff
@@ -131,6 +143,7 @@ func run(ctx context.Context, cfg cmds.Agent, proxy proxy.Proxy) error {
 	systemd.SdNotify(true, "READY=1\n")
 
 	<-ctx.Done()
+	wg.Wait()
```
Comment (manuelbuil) [inline pkg/agent/run.go:146]: Imagine something goes wrong in netpol and we get `ctx.Done()`. Could it be that flannel still runs and never executes `wg.Done()` and thus we stay stuck here?

Comment (vadorovsky) [reply pkg/agent/run.go:146]: Yes, that might be the case. I'm wondering if we should pass the `stopChan` to all goroutines, including the one with flannel and close that channel when after `<-ctx.Done` and then `wg.Wait()`. WDYT?

That would mean killing the whole agent when one component dies, which seems like was never done, but I think that's a good way of handling failures.

Or if we don't want to end k3s execution because of one component - we would need to come up with some recovery logic.

Comment (manuelbuil) [reply pkg/agent/run.go:146]: The problem I see is that the code of flannel with backend vxlan is not executing any `<-ctx.Done()`, so even if ctx finishes, I don't think the goroutine has a way to "exit" properly

**Comment by dweomer** [comment]
@vadorovsky can you take a look at before and after on the binary size for `dist/artifacts/k3s`? In discussing a change like this with @manuelbuil yesterday he did point out a legitimate concern that maybe the binary gets bigger which seems to be the case:
```
+ stat dist/artifacts/k3s
  File: dist/artifacts/k3s
  Size: 53260288        Blocks: 104024     IO Block: 4096   regular file
Device: 85h/133d        Inode: 10378572    Links: 1
Access: (0755/-rwxr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
```

This is probably fine but with changes like this I would like to see the delta.

File: pkg/agent/run.go (hunk 2 above)
Comment (vadorovsky) [reply pkg/agent/run.go:146]: That doesn't sound good. I also see that in case of flannel failure we just log it without any other reaction. 

https://github.com/k3s-io/k3s/blob/5acd0b9008c93e9ce64e7b7beeabda9248408374/pkg/agent/flannel/setup.go#L100-L105

Should we somehow change it?

**Comment by vadorovsky** [comment]
@dweomer 

```
  master → origin U:1  ~/repos/k3s                                                                                                                                                                            18:12:28  vadorovsky 
❯ stat dist/artifacts/k3s
  File: dist/artifacts/k3s
  Size: 51650560  	Blocks: 100880     IO Block: 4096   regular file
Device: 0,38	Inode: 6015394     Links: 1
Access: (0755/-rwxr-xr-x)  Uid: ( 1000/vadorovsky)   Gid: (  100/   users)
Access: 2021-11-18 18:11:40.000000000 +0100
Modify: 2021-11-18 18:11:40.000000000 +0100
Change: 2021-11-18 18:12:25.726433631 +0100
 Birth: 2021-11-18 18:12:25.698433679 +0100
```
```
  use-kr-as-lib U:1  ~/repos/k3s                                                                                                                                                                              18:15:15  vadorovsky 
❯ stat dist/artifacts/k3s
  File: dist/artifacts/k3s
  Size: 51658752  	Blocks: 100896     IO Block: 4096   regular file
Device: 0,38	Inode: 6016096     Links: 1
Access: (0755/-rwxr-xr-x)  Uid: ( 1000/vadorovsky)   Gid: (  100/   users)
Access: 2021-11-18 18:14:24.000000000 +0100
Modify: 2021-11-18 18:14:24.000000000 +0100
Change: 2021-11-18 18:15:11.502156972 +0100
 Birth: 2021-11-18 18:15:11.474157019 +0100
```

To sum it up:

* master: `51650560`
* this PR: `51658752`
* delta: `8192`

File: pkg/agent/run.go (hunk 2 above)
Comment (brandond) [reply pkg/agent/run.go:146]: This isn't a log-and-ignore; `Fatalf()` will cause the k3s process to exit.

**Review by brandond — APPROVED** [review]

File: pkg/agent/run.go (hunk 2 above)
Comment (vadorovsky) [reply pkg/agent/run.go:146]: So you would be fine with the current code, but without `wg.Wait()`?

Comment (vadorovsky) [reply pkg/agent/run.go:146]: @brandond ping, please let me know what you think ^^

File: pkg/agent/run.go [hunk 3]
```diff
@@ -43,7 +45,11 @@ import (
 	utilpointer "k8s.io/utils/pointer"
 )
 
+const healthControllerChannelLength = 10
+
 func run(ctx context.Context, cfg cmds.Agent, proxy proxy.Proxy) error {
+	var wg sync.WaitGroup
```
Comment (brandond) [inline pkg/agent/run.go:51]: We're still not doing anything with the waitgroup; if we're not waiting on it (which I don't think we want to do) then lets remove it from the flannel.Run and netpol.Run signatures and allow them to continue to make their own as needed?

File: pkg/agent/run.go (hunk 1 above)
Comment (brandond) [reply pkg/agent/run.go:135]: I think stopChan should come from `ctx.Done()` so that it's plumbed through?  Not sure about the healthChan though, as discussed during standup.

**Review by brandond — CHANGES_REQUESTED** [review]

File: pkg/agent/run.go (hunk 3 above)
Comment (vadorovsky) [reply pkg/agent/run.go]: Alright, I moved the WaitGroup only to netpol.go, since it's impossible to run kube-router without passing `wg`. Then I'm not doing anything with it.

**Comment by vadorovsky** [comment]
Marking again as a draft. Since we don't want to support multiple/alternative CNIs in k3s and we want to support dual-stack in kube-router (from what @manuelbuil says), we will most likely end up forking kube-router, adding DS patches there and then vendoring the fork.

**Comment by vadorovsky** [comment]
New PR #5079
